### ISSUE TITLE
add lint for empty `vec!` invocation

### ISSUE THREAD

**Issue by andrewbanchich** [rust-lang/rust-clippy#12774]
### What it does

instead of calling `vec![]` to create an empty vec, use `Vec::new()`. the `vec!` macro's purpose is for adding any number of items to a vec in an ergonomic and optimized way, so there is no advantage to creating an empty vec with a macro.

`vec![]` will expand to `Vec::new()` anyway, so just write `Vec::new()`.

### Advantage

- remove unnecessary macro call for times you just need an empty vec
-- not sure if this counts as a micro/nano-optimization since there no more macro expansion
-- at the very least, i think it's good to reserve macros for only when they provide some value

### Example

```rust
let nums: Vec<usize> = vec![];
```

Could be written as:

```rust
let nums: Vec<usize> = Vec::new();
```

**Comment by matthiaskrgr** [comment]
Note that there have been cases where the macro actually generated better code than vec::new()

https://github.com/rust-lang/rust/issues/54475

**Comment by Nilstrieb** [comment]
I don't think this should be more than a restriction lint. There is nothing clearer and more readable about new imo, especially when there are several other uses of the vec macro with values in the same function, it's more consistent.
